components/script/dom/vr.rs

```diff
@@ -128,34 +131,112 @@ impl VR {
-        match &event {
-            &webvr::VRDisplayEvent::Connect(ref display) => {
-                let display = self.sync_display(&display);
-                display.handle_webvr_event(&event);
-                self.notify_event(&display, &event);
-            },
-            &webvr::VRDisplayEvent::Disconnect(id) => {
-                if let Some(display) = self.find_display(id) {
-                    display.handle_webvr_event(&event);
-                    self.notify_event(&display, &event);
-                }
+        match event {
+            WebVREventMsg::DisplayEvent(event) => {
+                match &event {
+                    &WebVRDisplayEvent::Connect(ref display) => {
+                        let display = self.sync_display(&display);
+                        display.handle_webvr_event(&event);
+                        self.notify_event(&display, &event);
+                    },
+                    &WebVRDisplayEvent::Disconnect(id) => {
+                        if let Some(display) = self.find_display(id) {
+                            display.handle_webvr_event(&event);
+                        self.notify_event(&display, &event);
+                        }
+                    },
+                    &WebVRDisplayEvent::Activate(ref display, _) |
+                    &WebVRDisplayEvent::Deactivate(ref display, _) |
+                    &WebVRDisplayEvent::Blur(ref display) |
+                    &WebVRDisplayEvent::Focus(ref display) |
+                    &WebVRDisplayEvent::PresentChange(ref display, _) |
+                    &WebVRDisplayEvent::Change(ref display) => {
+                        let display = self.sync_display(&display);
+                        display.handle_webvr_event(&event);
+                    }
+                };
             },
-            &webvr::VRDisplayEvent::Activate(ref display, _) |
-            &webvr::VRDisplayEvent::Deactivate(ref display, _) |
-            &webvr::VRDisplayEvent::Blur(ref display) |
-            &webvr::VRDisplayEvent::Focus(ref display) |
-            &webvr::VRDisplayEvent::PresentChange(ref display, _) |
-            &webvr::VRDisplayEvent::Change(ref display) => {
-                let display = self.sync_display(&display);
-                display.handle_webvr_event(&event);
+            WebVREventMsg::GamepadEvent(event) => {
+                match &event {
+                    &WebVRGamepadEvent::Connect(ref data, ref state) => {
+                        if let Some(gamepad) = self.find_gamepad(state.gamepad_id) {
+                            gamepad.update_from_vr(&state);
+                        } else {
+                            // new gamepad
+                            self.sync_gamepad(Some(data.clone()), state);
+                        }
+                    },
+                    &WebVRGamepadEvent::Disconnect(id) => {
+                        if let Some(gamepad) = self.find_gamepad(id) {
```

Comment:
Yes, it can happen in new "tabs"/JSContexts when the gamepads where already connected in the previous pages.